### PR TITLE
Add new build-tools update technique - build_update.sh

### PR DESCRIPTION
**Pull request by rfay**
Issue: https://github.com/drud/general/issues/17 (Overall build system for drud projects)

The subtree merge technique did not work in practice. The build_update.sh script here can do an add or update and commit the results.

### CODE REVIEW

File: build_update.sh [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+fi
+
+
+wget -q -O $local_file $tarball_url
+tar -xf $local_file
+mv $internal_name build-tools
+git add build-tools
+echo "Updated build-tools to $tag
+
+$base_url" | git commit -F -
```
Comment (beeradb) [inline build_update.sh]: I'm not sure if I love the commit here or not. I can definitely see an argument for not doing it.

**Comment by beeradb** [comment]
I possibly misunderstand the instructions but I had issues running this.

Steps to reproduce:

- check out repository without the build tools added (in this case I used jobwatcher)
- Create a `build-tools` directory.
- Add the build_update.sh script to that directory. Ensure it is executable.
- Change to that directory, and run `./build_update.sh`

Result:

```
jobwatcher/build-tools ‹gke:bbowman› git:(20170120_upgrade_kubectl) » pwd
/Users/beeradb/Projects/newmedia/jobwatcher/build-tools
jobwatcher/build-tools ‹gke:bbowman› git:(20170120_upgrade_kubectl) » ./build_update.sh
OK to replace current build-tools at /Users/beeradb/Projects/newmedia/jobwatcher/build-tools?
Replace build-tools with latest version? y/Ny
fatal: pathspec 'build-tools' did not match any files
```

File: build_update.sh (hunk 1 above)
Comment (rfay) [inline build_update.sh]: I imagine everybody knows how to `git reset --hard HEAD^`

**Comment by rfay** [comment]
No, you weren't supposed to create the directory for that scenario.

**Comment by beeradb** [comment]
After trying it again, it does seem to work.

I was a little bit confused about _how_ it worked, though, which was to pull stuff into a `jobwatcher/build-tools/build-tools-1.0` directory. Is the idea that you would update to the new version, then remove the previous folder? Is there any use case for running multiple versions side by side?

If we're just putting them in a folder to make it clear which version of the tools we have, perhaps always using just the `build-tools` directory, but adding a version.txt to the release would achieve the same thing?

Walking me through the thought process/reasoning behind the versioned folder would be immensely helpful.

**Comment by rfay** [comment]
We definitely need to share screens because I imagine you're finding something I can't replicate. I just added it to jobwatcher and saw this, not a versioned folder. (Note that the release tarball from github comes with a versioned folder inside it, which is why the script renames it to build-tools.)

Here's what I see when adding to jobwatcher for the first time:
```
rfay@Randy-Air:~/go/src/github.com/drud/jobwatcher$ ./build_update.sh
OK to add current build-tools at /Users/rfay/go/src/github.com/drud/jobwatcher/build-tools?
Add build-tools with latest version? y/Ny
[20170210_add_build_tools 0d8c4d9] Updated build-tools to 1.0
 13 files changed, 368 insertions(+)
 create mode 100644 build-tools/Makefile.example
 create mode 100644 build-tools/README.md
 create mode 100755 build-tools/build-scripts/build_go.sh
 create mode 100755 build-tools/build-scripts/test_go.sh
 create mode 100644 build-tools/circle.yml.example
 create mode 100644 build-tools/gitignore.example
 create mode 100644 build-tools/makefile_components/README.md
 create mode 100644 build-tools/makefile_components/base_build_go.mak
 create mode 100644 build-tools/makefile_components/base_build_python-docker.mak
 create mode 100644 build-tools/makefile_components/base_container.mak
 create mode 100644 build-tools/makefile_components/base_push.mak
 create mode 100644 build-tools/makefile_components/base_test_go.mak
 create mode 100644 build-tools/makefile_components/base_test_python.mak
```

I do think there's one thing that I probably have to fix: This determines the difference between an upgrade and an install by whether there is a build-tools directory in the parent directory. That fails if you have build-tools checked out (because it thinks it's doing an upgrade. I'll improve that check to say "Am I *in* the build-tools directory"

**Comment by rfay** [comment]
I was able to reproduce the versioned-folder problem you discovered, and am working on it. Definitely wrong.

**Comment by rfay** [comment]
The updated version I think supports all the excellent permutations you explored. It handles a manually created build-tools directory, a sibling build-tools directory, etc. Overall more robust.

File: build_update.sh [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+set -e
+
+base_url=$(curl -s -I https://github.com/drud/build-tools/releases/latest | awk '/^Location/ {gsub(/[\n\r]/,"",$2);  printf "%s", $2}')
+tag=${base_url##*/}
+
+tarball_url="https://github.com/drud/build-tools/archive/$tag.tar.gz"
+internal_name=build-tools-$tag
+local_file=/tmp/$internal_name.tgz
+
+set -x
```
Comment (beeradb) [inline build_update.sh]: [minor]: I don't believe we'll want this merged, correct?

**Comment by rfay** [comment]
Thanks for catching that - pushed update.
I'd like to tag a new release after https://github.com/drud/build-tools/pull/3 goes in.

**Pull request merged**